scheduler/src/cook/mesos.clj

```diff
@@ -156,38 +195,37 @@
+                                           :fenzo-fitness-calculator fenzo-fitness-calculator
+                                           :fenzo-floor-iterations-before-reset fenzo-floor-iterations-before-reset
+                                           :fenzo-floor-iterations-before-warn fenzo-floor-iterations-before-warn
+                                           :fenzo-max-jobs-considered fenzo-max-jobs-considered
+                                           :fenzo-scaleback fenzo-scaleback
+                                           :good-enough-fitness good-enough-fitness
+                                           :mea-culpa-failure-limit mea-culpa-failure-limit
+                                           :mesos-run-as-user mesos-run-as-user
+                                           :agent-attributes-cache agent-attributes-cache
+                                           :offer-incubate-time-ms offer-incubate-time-ms
+                                           :pool-name->pending-jobs-atom pool-name->pending-jobs-atom
+                                           :rebalancer-reservation-atom rebalancer-reservation-atom
+                                           :task-constraints task-constraints
+                                           :trigger-chans trigger-chans})
                                         cluster-connected-chans (->> cluster-name->compute-cluster
                                                                      (map (fn [[compute-cluster-name compute-cluster]]
                                                                             (try
                                                                               (cc/initialize-cluster compute-cluster
-                                                                                                     pool-name->fenzo
-                                                                                                     running-tasks-ents)
+                                                                                                     pool-name->fenzo)
                                                                               (catch Throwable t
                                                                                 (log/error t "Error launching compute cluster" compute-cluster-name)
                                                                                 ; Return a chan that never gets a message on it.
                                                                                 (async/chan 1)))))
                                                                      ; Note: This doall has a critical side effect of actually initializing
                                                                      ; all of the clusters.
                                                                      doall)]
+                                    (reset! cc/pool-name->fenzo-atom (:pool-name->fenzo scheduler))
```

Comment:
we don't create any dynamic clusters at this point. that happens below at line 262

Reply:
Well, you'll want to make sure this is contained within the body of the lock you add in. We want to make sure it doesn't get accidently moved outside of the lock. (I suggest a comment to that effect in either case, this code must run when the lock is held)

Reply:
why does this need to be locked? it doesn't change. could also just be global.

Reply:
The race is REST cluster creation. We want to make sure they arrive after we've finished initialization, including this line, Putting this in the lock guarantees that, because REST queries grab the lock before doing anything.

Reply:
which lock do REST queries grab?
But also, we should not have a REST cluster creation query until we assume leadership. right?

Reply:
actually, I moved create-datomic-scheduler before setting cc/pool-name->fenzo-atom and before (reset! leadership-atom true). This way cc/pool-name->fenzo-atom is ready before any REST calls are made to it